### PR TITLE
Add support for bugsnagReporting in gradle config.

### PR DESCRIPTION
**Pull request by tabletenniser**
See https://github.com/MarathonLabs/marathon/issues/818. This PR essentially adds feature parity between `analyticsTracking` and `bugsnagReporting`

This PR also:
* Fix a bug where we still track analytics even if `analyticsTracking` is set to False in the gradle config
* Fix an incorrect CLI helper message on analyticsTracking where the default value is now to track analytics

Things are a bit messy today because `analyticsTracking` can be set in both Marathonfile / gradle config and CLI today. In the case where we run Marathon from gradle, we can only set it inside gradle config but not the CLI. Before this PR, `analyticsTracking` defaults to False for Marathon config but defaults to True for CLI. This makes it impossible to disable `analyticsTracking` when running Marathon from Gradle because the CLI's default True value will always take priority and overwrite whatever value we have in the Marathonfile / gradle config. This PR changes it such that we default to True for both Marathonfile / gradle config and CLI, and if we set it to False explicitly in either, we will disable the analytics tracking. A similar pattern is also applied to `bugsnagReporting`.

This PR is tested by:
* End to end using the `samples/android-app` sample app where we see `Noop exceptions reporter started` with `bugsnagReporting` set to False in the gradle config.
* Unit tests on `ConfigurationFactory.kt`

### CODE REVIEW

File: cli/src/main/kotlin/com/malinskiy/marathon/cli/ApplicationView.kt
Comment (Malinskiy) [inline cli/src/main/kotlin/com/malinskiy/marathon/cli/ApplicationView.kt:48]: Any crash reporting should start at the beginning of the code to check as many problems as possible. One example was the classpath incompatibility between gradle's jackson and marathon's jackson which failed to even parse config. This issue is not applicable anymore, but the reasoning for it still applies.

File: configuration/src/main/kotlin/com/malinskiy/marathon/config/Configuration.kt
Comment (Malinskiy) [inline configuration/src/main/kotlin/com/malinskiy/marathon/config/Configuration.kt:113]: Good catch

File: core/src/main/kotlin/com/malinskiy/marathon/Marathon.kt
Comment (Malinskiy) [inline core/src/main/kotlin/com/malinskiy/marathon/Marathon.kt]: How is this relevant to the PR?

File: marathon-gradle-plugin/src/main/kotlin/com/malinskiy/marathon/gradle/MarathonExtension.kt
Comment (Malinskiy) [inline marathon-gradle-plugin/src/main/kotlin/com/malinskiy/marathon/gradle/MarathonExtension.kt:227]: :+1:

File: sample/android-app/app/build.gradle.kts
Comment (Malinskiy) [inline sample/android-app/app/build.gradle.kts]: Please remove these from the sample app. Instead, maybe the docs need an example of disabling bugsnag for both CLI and for Gradle, currently there are none.

File: configuration/src/main/kotlin/com/malinskiy/marathon/config/serialization/ConfigurationFactory.kt
Comment (Malinskiy) [inline configuration/src/main/kotlin/com/malinskiy/marathon/config/serialization/ConfigurationFactory.kt]: might be easier to rewrite with De Morgan's law into 
`analyticsTracking == true && configuration.analyticsTracking == true`, same below

**Review by Malinskiy — CHANGES_REQUESTED** [review]

File: cli/src/main/kotlin/com/malinskiy/marathon/cli/ApplicationView.kt
Comment (Malinskiy) [inline cli/src/main/kotlin/com/malinskiy/marathon/cli/ApplicationView.kt:94]: Generally if there is a usage of Noop implementation null checks are redundant

File: cli/src/main/kotlin/com/malinskiy/marathon/cli/ApplicationView.kt
Comment (tabletenniser) [reply cli/src/main/kotlin/com/malinskiy/marathon/cli/ApplicationView.kt:48]: I think it's a chicken-and-egg problem here. This PR adds the ability to disable bugSnag reporting from the gradle / Marathonfile. But without parsing the Marathon file, we don't know if the setting is disabled.

I guess the best we can do is to always start bugSnag reporting, and if it's disabled in the Marathon file, we terminate it right away. What do you think?

File: cli/src/main/kotlin/com/malinskiy/marathon/cli/ApplicationView.kt
Comment (tabletenniser) [reply cli/src/main/kotlin/com/malinskiy/marathon/cli/ApplicationView.kt:94]: This `if condition` is added in case we fail to parse the Marathon file and throw an exception. In this case, `bugsnagExceptionsReporter` isn't initialized yet and we don't want to call `bugsnagExceptionsReporter.end()`

File: configuration/src/main/kotlin/com/malinskiy/marathon/config/serialization/ConfigurationFactory.kt
Comment (tabletenniser) [reply configuration/src/main/kotlin/com/malinskiy/marathon/config/serialization/ConfigurationFactory.kt]: Actually, I don't think it's quite the same because `null` value is allowed for `analyticsTracking`. If `analyticsTracking` is null and `configuration.analyticsTracking` is true, we want it to be enabled (which is what the code currently does). We want to disable it only if one *explicitly* sets it to false for `analyticsTracking` or `configuration.analyticsTracking`.

Though based on my reading of the code, `null` value seems to only be used in unit tests?  Maybe we can make it a mandatory param and dis-allow `null` values?

File: core/src/main/kotlin/com/malinskiy/marathon/Marathon.kt
Comment (tabletenniser) [reply core/src/main/kotlin/com/malinskiy/marathon/Marathon.kt]: Oops, that was not intentional, added it back.

File: sample/android-app/app/build.gradle.kts
Comment (tabletenniser) [reply sample/android-app/app/build.gradle.kts]: Removed and updated `docs/docs/intro/configure.md`

File: docs/docs/intro/configure.md
Comment (Malinskiy) [inline docs/docs/intro/configure.md]: Can you please just use https://docusaurus.io/docs/markdown-features/admonitions?
There are examples in the codebase for them already

File: docs/docs/intro/configure.md
Comment (Malinskiy) [inline docs/docs/intro/configure.md]: Same with admonitions

**Review by Malinskiy — CHANGES_REQUESTED** [review]

**Review by Malinskiy — DISMISSED** [review]

**Comment by tabletenniser** [comment]
Thanks @Malinskiy for the review! Are you going to merge this PR (since I obviously don't have permission to) or do we still need reviews from @tagantroy and @Vacxe ?

**Comment by Malinskiy** [comment]
I'm waiting for one more lgtm for a day and will merge afterwards assuming no issues.

File: cli/src/main/kotlin/com/malinskiy/marathon/cli/ApplicationView.kt
Comment (matzuk) [inline cli/src/main/kotlin/com/malinskiy/marathon/cli/ApplicationView.kt:49]: `lateinit var`?

File: configuration/src/main/kotlin/com/malinskiy/marathon/config/serialization/ConfigurationFactory.kt
Comment (matzuk) [inline configuration/src/main/kotlin/com/malinskiy/marathon/config/serialization/ConfigurationFactory.kt]: To be honest, such constructions are a bit confusing. For me, it is better to write a bit more but make the statement clear. For example,
```
analyticsTracking = 
    if (analyticsTracking == false || configuration.analyticsTracking == false) false 
    then true
```
But this comment may be subjective.

**Review by matzuk — DISMISSED** [review]

File: configuration/src/main/kotlin/com/malinskiy/marathon/config/serialization/ConfigurationFactory.kt
Comment (matzuk) [reply configuration/src/main/kotlin/com/malinskiy/marathon/config/serialization/ConfigurationFactory.kt]: @tabletenniser thank you for your contribution! Approved.
I've added a small comment. Please have a look.

File: configuration/src/main/kotlin/com/malinskiy/marathon/config/serialization/ConfigurationFactory.kt
Comment (matzuk) [inline configuration/src/main/kotlin/com/malinskiy/marathon/config/serialization/ConfigurationFactory.kt]: why `run` is used here? =)

Comment (tabletenniser) [reply configuration/src/main/kotlin/com/malinskiy/marathon/config/serialization/ConfigurationFactory.kt]: Because without it, I'm getting a compilation error: `Type mismatch: inferred type is () -> Boolean but Boolean was expected`  (I'm a completely newbie to Kotlin, so if there's a more proper way to fix the issue, feel free to suggest it!)

Comment (matzuk) [reply configuration/src/main/kotlin/com/malinskiy/marathon/config/serialization/ConfigurationFactory.kt]: No worries, I've fixed it for you.

**Review by matzuk — APPROVED** [review]

**Comment by matzuk** [comment]
@Malinskiy let's merge

**Pull request merged**
